### PR TITLE
[MLOPS-626] Adding `limits`-argument to `list` methods. 

### PR DESCRIPTION
**Pull request by qTipTip**
This PR introduces the possibility of specifying a `limit` in the `list`-methods of `FunctionsAPI`, `FunctionsCallAPI`, and `FunctionsScheduleAPI`. If not specified, the `limit`-default is controlled by `LIST_LIMIT_DEFAULT` and is set to 25.

### CODE REVIEW

**Comment by qTipTip** [comment]
Is a default limit of 25 sensible? It's what Assets has.

File: cognite/experimental/_api/functions.py [hunk 1]
```diff
@@ -162,9 +164,12 @@ def delete(self, id: Union[int, List[int]] = None, external_id: Union[str, List[
         """
         self._delete_multiple(ids=id, external_ids=external_id, wrap_ids=True)
 
-    def list(self) -> FunctionList:
+    def list(self, limit: Optional[int] = LIST_LIMIT_DEFAULT) -> FunctionList:
         """`List all functions. <https://docs.cognite.com/api/playground/#operation/get-function>`_
 
+        Args:
+            limit (int, optional): Maximum number of functions to list. Pass in -1, 'inf' or None to list all functions.
```
Comment (HMEiding) [inline cognite/experimental/_api/functions.py:171]: Should we be consistent and say `float("inf")` like elsewhere?

File: cognite/experimental/_api/functions.py [hunk 2]
```diff
@@ -177,7 +182,12 @@ def list(self) -> FunctionList:
                 >>> functions_list = c.functions.list()
         """
         url = "/functions"
-        res = self._get(url)
+
+        if limit in ["inf", -1]:
+            limit = None
```
Comment (HMEiding) [inline cognite/experimental/_api/functions.py:187]: We agreed that this should be all elements, right? Passing `None` will result in the default limit in the API.

File: cognite/experimental/_api/functions.py [hunk 3]
```diff
@@ -397,10 +407,12 @@ def list(
         schedule_id: Optional[int] = None,
         start_time: Optional[Dict[str, int]] = None,
         end_time: Optional[Dict[str, int]] = None,
+        limit: Optional[int] = LIST_LIMIT_DEFAULT,
     ) -> FunctionCallList:
         """List all calls associated with a specific function id. Either function_id or function_external_id must be specified.
 
         Args:
+            limit (int, optional): Maximum number of function calls to list. Pass in -1, 'inf' or None to list all Function Calls.
```
Comment (HMEiding) [inline cognite/experimental/_api/functions.py:415]: The `self._list()` method requires `float("inf")` here:
https://github.com/cognitedata/cognite-sdk-python/blob/52645b05b00c81374a29cec412790067ae3af1a8/cognite/client/_api_client.py#L403

File: cognite/experimental/_api/functions.py [hunk 4]
```diff
@@ -568,9 +579,12 @@ def retrieve(self, id: int) -> Optional[FunctionSchedule]:
         utils._auxiliary.assert_exactly_one_of_id_or_external_id(id=id, external_id=None)
         return self._retrieve_multiple(ids=id, wrap_ids=True)
 
-    def list(self) -> FunctionSchedulesList:
+    def list(self, limit: Optional[int] = LIST_LIMIT_DEFAULT) -> FunctionSchedulesList:
         """`List all schedules associated with a specific project. <https://docs.cognite.com/api/playground/#operation/get-api-playground-projects-project-functions-schedules>`_
 
+        Args:
+            limit (int, optional): Maximum number of schedules to list. Pass in -1, 'inf' or None to list all schedules.
```
Comment (HMEiding) [inline cognite/experimental/_api/functions.py:586]: Should we be consistent and say `float("inf")` like elsewhere?

File: cognite/experimental/_api/functions.py [hunk 5]
```diff
@@ -591,7 +605,12 @@ def list(self) -> FunctionSchedulesList:
 
         """
         url = f"/functions/schedules"
-        res = self._get(url)
+
+        if limit in ["inf", -1]:
+            limit = None
```
Comment (HMEiding) [inline cognite/experimental/_api/functions.py:610]: We agreed that this should be all elements, right? Passing `None` will result in the default limit in the API.

File: tests/tests_unit/test_api/test_functions.py [hunk 6]
```diff
@@ -225,6 +236,17 @@ def mock_function_calls_list_response(rsps):
     yield rsps
 
 
+@pytest.fixture
+def mock_function_calls_list_response_with_limit(rsps):
```
Comment (HMEiding) [inline tests/tests_unit/test_api/test_functions.py:240]: Nitpick, but the `/list` endpoint is typically referred to as the "filter"-endpoint:
https://docs.cognite.com/api/v1/#operation/listAssets

I don't think we've been consistent with this, but we can try. 😆

File: tests/tests_unit/test_api/test_functions.py [hunk 7]
```diff
@@ -225,6 +236,17 @@ def mock_function_calls_list_response(rsps):
...
+    response_body = {"items": [CALL_COMPLETED, CALL_SCHEDULED]}
```
Comment (HMEiding) [inline tests/tests_unit/test_api/test_functions.py:241]: FYI: The response body also returns a cursor, although I guess the tests don't operate on it. I like it to be as realistic as possible, but maybe we could skip it here. What do you think? Have you checked assets?

File: tests/tests_unit/test_api/test_functions.py [hunk 8]
```diff
@@ -225,6 +236,17 @@ def mock_function_calls_list_response(rsps):
...
+    limit = 2
+    query_parameters = f"?limit={limit}"
```
Comment (HMEiding) [inline tests/tests_unit/test_api/test_functions.py:243]: We don't have query parameters to this endpoint.

**Review by HMEiding — CHANGES_REQUESTED** [review]
Great! Just a few small comments.

File: cognite/experimental/_api/functions.py (hunk 1 above)
Comment (qTipTip) [reply cognite/experimental/_api/functions.py:171]: Done

File: cognite/experimental/_api/functions.py (hunk 2 above)
Comment (qTipTip) [reply cognite/experimental/_api/functions.py:187]: Done

File: cognite/experimental/_api/functions.py (hunk 3 above)
Comment (qTipTip) [reply cognite/experimental/_api/functions.py:415]: Done

File: cognite/experimental/_api/functions.py (hunk 4 above)
Comment (qTipTip) [reply cognite/experimental/_api/functions.py:586]: Done

File: cognite/experimental/_api/functions.py (hunk 5 above)
Comment (qTipTip) [reply cognite/experimental/_api/functions.py:610]: Done

File: tests/tests_unit/test_api/test_functions.py (hunk 6 above)
Comment (qTipTip) [reply tests/tests_unit/test_api/test_functions.py:240]: Haha, ok :) Might as well try!

File: tests/tests_unit/test_api/test_functions.py (hunk 8 above)
Comment (qTipTip) [reply tests/tests_unit/test_api/test_functions.py:243]: True! Was left from initial testing.

File: tests/tests_unit/test_api/test_functions.py (hunk 6 above)
Comment (qTipTip) [reply tests/tests_unit/test_api/test_functions.py:240]: Should I refactor the above one as well then? mock_function_calls_list_response

File: tests/tests_unit/test_api/test_functions.py (hunk 7 above)
Comment (qTipTip) [reply tests/tests_unit/test_api/test_functions.py:241]: Assets have no reference to cursors in their tests. 

Not sure how crucial it is to mock this here.

**Review by HMEiding — APPROVED** [review]
LGTM 💯

File: cognite/experimental/_api/functions.py [hunk 9]
```diff
@@ -177,7 +182,12 @@ def list(self) -> FunctionList:
                 >>> functions_list = c.functions.list()
         """
         url = "/functions"
-        res = self._get(url)
+
+        if limit in [float("inf"), -1, None]:
```
Comment (matzhaugen) [inline cognite/experimental/_api/functions.py:186]: The type of limit is `Optional[int]`, do you need to check `float('"inf")`?

File: cognite/experimental/_api/functions.py [hunk 10]
```diff
@@ -591,7 +605,12 @@ def list(self) -> FunctionSchedulesList:
 
         """
         url = f"/functions/schedules"
-        res = self._get(url)
+
+        if limit in [float("inf"), -1, None]:
```
Comment (matzhaugen) [inline cognite/experimental/_api/functions.py:609]: Same as above

**Review by matzhaugen — CHANGES_REQUESTED** [review]

**Comment by matzhaugen** [comment]
Small comment, otherwise good.

**Review by matzhaugen — APPROVED** [review]

**Pull request merged**